File: lib/status-transitions.ts
import { User } from './use-user-management';

const transitions = {
  'Front Desk': {
    'Completed': ['Ready for Pickup'],
    'Ready for Pickup': ['Picked Up'],
    'Pending': ['Cancelled'],
    'In Progress': ['Cancelled'],
    'Awaiting Parts': ['Cancelled'],
  },
  'Technician': {
    'Pending': ['In Progress'],
    'In Progress': ['Awaiting Parts', 'Completed'],
    'Awaiting Parts': ['In Progress'],
  },
  'Manager': {
  },
};

export const getAllowedStatusTransitions = (role: User["role"], currentStatus: string): string[] => {
  if (role === 'Manager') {
    // Managers can transition to any status
    return Object.values(transitions).reduce((acc, val) => acc.concat(Object.keys(val)), []);
  }

  const roleTransitions = transitions[role];
  if (!roleTransitions) {
    return [];
  }

  return roleTransitions[currentStatus] || [];
};

File: django_backend/Eapp/status_transitions.py

from .models import User, Task

ALLOWED_TRANSITIONS = {
    'Front Desk': {
        'Completed': ['Ready for Pickup', 'In Progress', 'Pending'],
        'Ready for Pickup': ['Picked Up', 'Pending', 'In Progress'],
        'Picked Up': ['In Progress'],
        'Pending': ['Terminated'],
        'In Progress': ['Terminated', 'Pending'],
    },
    'Technician': {
        'Pending': ['In Progress'],
        'In Progress': ['Awaiting Parts', 'Completed'],
        'Awaiting Parts': ['In Progress'],
    },
    'Manager': {
        # Managers can transition from any status to any other status.
    }
}

def can_transition(user: User, task: Task, new_status: str) -> bool:
    """
    Check if a user has permission to transition a task to a new status.
    """
    if user.is_superuser or user.role == User.Role.MANAGER:
        return True

    role_transitions = ALLOWED_TRANSITIONS.get(user.role, {})
    current_status = task.status
    
    allowed_next_statuses = role_transitions.get(current_status, [])
    return new_status in allowed_next_statuses
